src/main/scala/mesosphere/marathon/experimental/repository/TemplateRepository.scala

```diff
@@ -0,0 +1,184 @@
+    * templates data to a byte array.
+    *
+    * @param template
+    * @return
+    */
+  def toNode(template: Template[_]): Node = Node(storePath(template.id, version(template)), ByteString(template.toProtoByteArray))
+
+  /**
+    * Convenience method that returns templates version by calling [[TemplateRepositoryLike.storePath]] method internally.
+    *
+    * @param template
+    * @return
+    */
+  def storePath(template: Template[_]): String = storePath(template.id, version(template))
+
+  /**
+    * Method does the opposite conversion to the [[toNode]] method by taking in a trie [[mesosphere.marathon.experimental.storage.PathTrie.TrieNode]]
+    * node (which might be `null` if the node doesn't exist) and a template instance. Returns a [[Success[T] ]] if data exists or a
+    * [[Failure[NoNodeException] ]] if it doesn't.
+    *
+    * @param maybeData
+    * @param template
+    * @tparam T
+    * @return
+    */
+  def toTemplate[T](maybeData: Array[Byte], template: Template[T]): Try[T] = maybeData match {
+    case null => Failure(new NoNodeException(s"Template with path ${template.id} not found"))
+    case data => Try(template.mergeFromProto(data))
+  }
+
+  /**
+    * Method recursively reads the repository structure without reading node's data.
+    * @param path start path. children below this path will be read recursively.
+    * @return
+    */
+  private[this] def tree(path: String): Future[Done] = async {
+    val Success(children) = await(store.children(path, absolute = true))
+    val grandchildren = children.map { child =>
+      trie.addPath(child)
+      tree(child)
+    }
+    await(Future.sequence(grandchildren).map(_ => Done))
+  }
+
+  /**
+    * Method fetches the data for repo's leaf nodes since that is where we keep the templates.
+    * @return
+    */
+  private[this] def data(): Future[Done] = {
+    Source(trie.getLeafs("/").asScala.toList)
+      .via(store.readFlow)
+      .map {
+        case Success(node) => trie.addPath(node.path, node.data.toArray)
+        case Failure(ex) => throw new IllegalStateException("Failed to initialize template repository. " +
+          "Apparently one of the nodes was deleted during initialization.", ex)
+      }
+      .runWith(Sink.ignore)
+  }
+
+  /**
+    * Method initializes the repository by reading all the storage content and saving it in the path trie. First,
+    * tree structure is read, then, leaf nodes data is fetched. This method should be called prior to any other
+    * repository method call.
+    *
+    * @return
+    */
+  def initialize(): Future[Done] = async {
+    await(store.createIfAbsent(Node(base, ByteString.empty))) // Create the base path if it doesn't exist
+    await(tree(base))
+    await(data())
+  }
+
+  override def create(template: Template[_]): Future[String] = {
+    store
+      .create(toNode(template))
+      .map(_ => trie.addPath(storePath(template), template.toProtoByteArray))
+      .map(_ => version(template))
+  }
```

Comment:
I think we can avoid the creation of two futures with

```scala
  override def create(template: Template[_]): Future[String] = async {
    await(store.create(toNode(template)))
    trie.addPath(storePath(template), template.toProtoByteArray)
    version(template)
  }
```

This would also express that `trie.addPath` is not asynchronous.